lib/config.js

```diff
@@ -209,109 +149,313 @@ function Config(options) {
+        if (rootPath && !pathIsInside(path.dirname(localConfigFile), rootPath)) {
+            continue;
+        }
 
-        config = ConfigOps.merge(config, this.useSpecificConfig);
+        debug("Loading " + localConfigFile);
+        const localConfig = this.loadConfig(localConfigFile);
+
+        // Ignore empty config files
+        if (!localConfig) {
+            continue;
+        }
+
+        // Check for root flag
+        if (localConfig.root === true) {
+            rootPath = path.dirname(localConfigFile);
+        }
+
+        debug("Using " + localConfigFile);
+        configs.push(localConfig);
+        searched.push(localConfigDirectory);
     }
 
-    // Step 5: Merge in command line environments
-    debug("Merging command line environment settings");
-    config = ConfigOps.merge(config, { env: this.env });
 
-    // Step 6: Merge in command line rules
-    if (this.options.rules) {
-        debug("Merging command line rules");
-        config = ConfigOps.merge(config, { rules: this.options.rules });
+    if (!configs.length && !cache && !this.useSpecificConfig) {
+
+        // Fall back on the personal config from ~/.eslintrc
+        debug("Using personal config file");
+        const personalConfig = this.getPersonalConfig();
+
+        if (personalConfig) {
+            configs.push(personalConfig);
+        } else if (!hasRules(this.options) && !this.options.baseConfig) {
+
+            // No config file, no manual configuration, and no rules, so error.
+            const noConfigError = new Error("No ESLint configuration found.");
+
+            noConfigError.messageTemplate = "no-config-found";
+            noConfigError.messageData = {
+                directory,
+                filesExamined: localConfigFiles
+            };
+
+            throw noConfigError;
+        }
     }
 
-    // Step 7: Merge in command line globals
-    config = ConfigOps.merge(config, { globals: this.globals });
+    // Merged with any cached portion
+    configs.reverse();
+    cache = cache ? cache.concat(configs) : configs;
 
-    // Only override parser if it is passed explicitly through the command line or if it's not
-    // defined yet (because the final object will at least have the parser key)
-    if (this.parser || !config.parser) {
-        config = ConfigOps.merge(config, {
-            parser: this.parser
-        });
+    // Set the caches for the parent directories
+    for (let i = 0; i < searched.length; i++) {
+        const localConfigDirectory = searched[i],
+            subCache = cache.slice(0, cache.length - i);
+
+        this.localHierarchyCache[localConfigDirectory] = subCache;
     }
 
-    if (this.parserOptions) {
-        config = ConfigOps.merge(config, {
-            parserOptions: this.parserOptions
-        });
+    return cache;
+};
+
+
+
+/**
+ * Get the vector of applicable configs from the hierarchy for a given file (glob matching occurs here).
```

Comment:
I tried to think of a better word for this but came up blank. I fell back to describing it in detail in comments in the JSDoc:
>A vector is an array of config file paths, each optionally followed by one or more numbers that correspond to the indices of nested config blocks within the config file's overrides section.